File: app/src/main/java/com/futsch1/medtimer/reminders/scheduling/ReminderScheduler.java
package com.futsch1.medtimer.reminders.scheduling;

import androidx.annotation.NonNull;

import com.futsch1.medtimer.ScheduledReminder;
import com.futsch1.medtimer.database.FullMedicine;
import com.futsch1.medtimer.database.Reminder;
import com.futsch1.medtimer.database.ReminderEvent;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.stream.Collectors;

public class ReminderScheduler {
    private final TimeAccess timeAccess;

    public ReminderScheduler(TimeAccess timeAccess) {
        this.timeAccess = timeAccess;
    }

    public List<ScheduledReminder> schedule(@NonNull List<FullMedicine> fullMedicineWithTagsAndReminders, @NonNull List<ReminderEvent> reminderEvents) {
        ArrayList<Reminder> reminders = getReminders(fullMedicineWithTagsAndReminders);
        ArrayList<ScheduledReminder> scheduledReminders = new ArrayList<>();

        for (Reminder reminder : reminders) {
            Scheduling scheduling = new SchedulingFactory().create(reminder, reminderEvents, this.timeAccess);
            Instant reminderScheduledTime = scheduling.getNextScheduledTime();

            if (reminderScheduledTime != null) {
                scheduledReminders.add(new ScheduledReminder(getMedicine(reminder, fullMedicineWithTagsAndReminders), reminder, reminderScheduledTime));
            }
        }

        scheduledReminders.sort(Comparator.comparing(ScheduledReminder::timestamp));

        return scheduledReminders;
    }

    @SuppressWarnings("java:S6204") // Stream.toList() not available in SDK version selected
    private ArrayList<Reminder> getReminders(List<FullMedicine> fullMedicineWithTagsAndReminders) {
        ArrayList<Reminder> reminders = new ArrayList<>();
        for (FullMedicine medicineWithReminder : fullMedicineWithTagsAndReminders
        ) {
            //noinspection SimplifyStreamApiCallChains
            reminders.addAll(medicineWithReminder.reminders.stream().filter(r -> r.active).collect(Collectors.toList()));
        }
        return reminders;
    }

    private FullMedicine getMedicine(Reminder reminder, List<FullMedicine> fullMedicineWithTagsAndReminders) {
        int medicineId = reminder.medicineRelId;

        Optional<FullMedicine> medicineOptional = fullMedicineWithTagsAndReminders.stream().filter(mwr -> mwr.medicine.medicineId == medicineId).findFirst();
        if (medicineOptional.isPresent()) {
            return medicineOptional.get();
        } else {
            throw new NoSuchElementException();
        }
    }

    public interface TimeAccess {
        ZoneId systemZone();

        LocalDate localDate();
    }

}

File: app/src/main/java/com/futsch1/medtimer/reminders/scheduling/ReminderScheduler.kt
package com.futsch1.medtimer.reminders.scheduling

import com.futsch1.medtimer.core.datastore.PreferencesDataSource
import com.futsch1.medtimer.core.domain.model.Medicine
import com.futsch1.medtimer.core.domain.model.ReminderEvent
import com.futsch1.medtimer.core.domain.model.ScheduledReminder
import com.futsch1.medtimer.reminders.TimeAccess

class ReminderScheduler(private val timeAccess: TimeAccess, private val dataSource: PreferencesDataSource) {
    fun schedule(medicines: List<Medicine>, reminderEvents: List<ReminderEvent>): List<ScheduledReminder> {
        val scheduledReminders = mutableListOf<ScheduledReminder>()

        for (medicine in medicines) {
            for (reminder in medicine.reminders) {
                if (!reminder.active) {
                    continue
                }

                val scheduling = SchedulingFactory().create(reminder, medicine, reminderEvents, timeAccess, dataSource)
                val reminderScheduledTime = scheduling.getNextScheduledTime()

                if (reminderScheduledTime != null) {
                    scheduledReminders.add(ScheduledReminder(medicine, reminder, reminderScheduledTime))
                }
            }
        }

        scheduledReminders.sortWith(Comparator.comparing(ScheduledReminder::timestamp))

        return scheduledReminders
    }

}
